pkg/frontend/querymiddleware/step_align.go

```diff
@@ -7,20 +7,76 @@ package querymiddleware
 
 import (
 	"context"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+	"github.com/grafana/mimir/pkg/util/validation"
 )
 
+type stepAlignMiddleware struct {
+	next     Handler
+	limits   Limits
+	resolver tenant.Resolver
+	logger   log.Logger
+	aligned  *prometheus.CounterVec
+}
+
 // newStepAlignMiddleware creates a middleware that aligns the start and end of request to the step to
-// improved the cacheability of the query results.
-func newStepAlignMiddleware() Middleware {
+// improve the cacheability of the query results based on per-tenant configuration.
+func newStepAlignMiddleware(limits Limits, resolver tenant.Resolver, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	aligned := promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_query_frontend_queries_step_aligned_total",
+		Help: "Number of queries whose start or end times have been adjusted to be step-aligned.",
+	}, []string{"user"})
+
 	return MiddlewareFunc(func(next Handler) Handler {
-		return HandlerFunc(func(ctx context.Context, r Request) (Response, error) {
-			start := (r.GetStart() / r.GetStep()) * r.GetStep()
-			end := (r.GetEnd() / r.GetStep()) * r.GetStep()
-			return next.Do(ctx, r.WithStartEnd(start, end))
-		})
+		return &stepAlignMiddleware{
+			next:     next,
+			limits:   limits,
+			resolver: resolver,
+			logger:   logger,
+			aligned:  aligned,
+		}
 	})
 }
 
+func (s *stepAlignMiddleware) Do(ctx context.Context, r Request) (Response, error) {
+	tenants, err := s.resolver.TenantIDs(ctx)
+	if err != nil {
+		return s.next.Do(ctx, r)
+	}
+
+	if validation.AllTrueBooleansPerTenant(tenants, s.limits.AlignQueriesWithStep) {
+		start := (r.GetStart() / r.GetStep()) * r.GetStep()
+		end := (r.GetEnd() / r.GetStep()) * r.GetStep()
+
+		if start != r.GetStart() || end != r.GetEnd() {
+			for _, id := range tenants {
+				s.aligned.WithLabelValues(id).Inc()
+			}
+
+			spanlog := spanlogger.FromContext(ctx, s.logger)
+			spanlog.DebugLog(
+				"msg", "query start or end has been adjusted to be step-aligned",
+				spanlogger.TenantIDsTagName, tenants,
```

Comment:
`spanlogger.New()` adds tenant IDs automatically but the middlewares package is a mix of lots of different ways to create spans so I'm adding it explicitly here because the `instrumentation` middleware uses `opentracing` directly.